File: AILang/benchmarks/c/slice_sum.c
#include <stdint.h>
#include <stdio.h>

typedef struct {
    int64_t *data;
    int64_t length;
} IntSlice;

static int64_t slice_sum_bench(int64_t iterations) {
    int64_t storage[8] = {3, 1, 4, 1, 5, 9, 2, 6};
    IntSlice view = {storage, 8};
    int64_t acc = 0;
    for (int64_t i = 0; i < iterations; ++i) {
        for (int64_t j = 0; j < view.length; ++j) {
            acc += view.data[j];
        }
    }
    return acc;
}

int main(void) {
    const int64_t iterations = 250000;
    const int64_t result = slice_sum_bench(iterations);
    printf("%lld\n", (long long)result);
    return 0;
}

File: AILang/benchmarks/rust/slice_sum.rs
fn slice_sum_bench(iterations: i64) -> i64 {
    let data: [i64; 8] = [3, 1, 4, 1, 5, 9, 2, 6];
    let view: &[i64] = &data;
    let mut acc: i64 = 0;
    for _ in 0..iterations {
        for &value in view {
            acc += value;
        }
    }
    acc
}

fn main() {
    let iterations: i64 = 250_000;
    let result = slice_sum_bench(iterations);
    println!("{result}");
}
